nomad/state/state_store_host_volumes.go

```diff
@@ -156,6 +157,30 @@ func (s *StateStore) HostVolumes(ws memdb.WatchSet, sort SortOption) (memdb.Resu
 	return s.hostVolumesIter(ws, indexID, sort)
 }
 
+// HostVolumesByIDPrefix retrieves all host volumes by ID prefix. Because the ID
+// index is namespaced, we need to handle the wildcard namespace here as well.
+func (s *StateStore) HostVolumesByIDPrefix(ws memdb.WatchSet, ns, prefix string, sort SortOption) (memdb.ResultIterator, error) {
+
+	if ns != structs.AllNamespacesSentinel {
+		return s.hostVolumesIter(ws, "id_prefix", sort, ns, prefix)
+	}
+
+	// for wildcard namespace, wrap the iterator in a filter function that
+	// filters all volumes by prefix
+	iter, err := s.hostVolumesIter(ws, indexID, sort)
+	if err != nil {
+		return nil, err
+	}
+	wrappedIter := memdb.NewFilterIterator(iter, func(raw any) bool {
+		vol, ok := raw.(*structs.HostVolume)
+		if !ok {
+			return true
+		}
+		return !strings.HasPrefix(vol.ID, prefix)
```

Comment:
this took me a sec to parse, so I'm typing it in english to make sure I grasp it:

https://pkg.go.dev/github.com/hashicorp/go-memdb#FilterFunc

> FilterFunc is a function that takes the results of an iterator and returns whether the result should be filtered out. 

`true` = filter it out (do not keep it)
`false` = do not filter it (keep it)

so,
`if !ok` - if it is not a `HostVolume` (somehow), `true` = throw it away
`if !strings.HasPrefix` - if it does not have the prefix, `true` = throw it away

Reply:
Right, it's "filter-out", not "filter-for". I feel like most of Nomad's code uses "filter-for" but this is coming from go-memdb where it's the opposite.